Rz endian docs

createyourpersonalaccount opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

I noticed that the https://github.com/rizinorg/rizin/blob/5211cdcfb06cc0740d568d54a2a160453a36869e/librz/include/rz_endian.h header is undocumented and decided to document the functions therein. I also inserted new `UT24_MAX` (and `MIN`) macros. I added some missing functions so that every endianness and bit combination exists (except for some middle-endian combinations which I wasn't sure about).  I also added a macro identifier in the preprocessor namespace in case the builtin swap functions did not exist for uniformity in behavior, and ability to document. A function parameter was renamed from `_dest` to `dest` to match what other functions call their parameter.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

createyourpersonalaccount committed:
librz: rename function parameter

The function parameter is renamed from _dest to dest. In the body of
the function, the identifiers _dest and dest are swapped
accordingly. This makes these functions have the same parameter name
as all others of the same kind in the header.

createyourpersonalaccount committed:
librz: add swap macros when builtins missing

The rz_swap_ macros are only defined when the compiler builtins are
available. We define them in case they are not for uniformity of
presence in the preprocessor namespace.

createyourpersonalaccount committed:
librz: add documentation for rz_endian.h

Every function and macro is documented except for the signed swap
macros.

createyourpersonalaccount committed:
librz: return max value in read if parameter is NULL

The read functions in rz_endian.h sometimes returned the MAX value
when the source parameter was NULL and sometimes didn't. For example,
compare rz_read_be16 which doesn't check NULL and rz_read_le16 which
does.

We fix this by inserting checks for NULL in every integer read
function.

We also insert these checks for the floating-point read
functions. They also too sometimes checked for NULL indirectly by
calling the corresponding integer read functions, but the read
functions with the offset parameter did not check for NULL.

Separately, The NULL checks in rz_read_le8 and rz_read_me8 were
removed because they were duplicate; they call rz_read_ble8 which
checks for NULL.

createyourpersonalaccount committed:
librz: add documentation in swap functions for NULL source parameter

The case where the source parameter is NULL is documented in all the
read functions in rz_endian.h

A mistake in the documentation of rz_read_at_ble64 is fixed; it
incorrectly mentioned 128-bit instead of 64-bit integers.

createyourpersonalaccount committed:
librz: add missing 24-bit macros

The 24-bit macros for minimum and maximum value are added.

createyourpersonalaccount committed:
librz: add missing endianness functions

For the readers and writers, certain endianness and bit number
combinations were missing. Some missing big-endian, little-endian,
middle-endian functions as well as helper functions are added.

The documentation of some read blue functions is corrected to mention
"specified" order instead of "big-endian".

createyourpersonalaccount committed:
librz: fixing some compiler errors from typos

ret2libc commented on librz/include/rz_endian.h:
```diff
@@ -242,27 +555,84 @@ static inline ut16 rz_read_at_le16(const void *src, size_t offset) {
+ * \attention If \p src is \c NULL then \c UT24_MAX is returned.
+ */
+static inline ut32 rz_read_le24(const void *src) {
+	if (!src) {
+		return UT24_MAX;
+	}
```
I wonder whether we should just return UT32_MAX in these cases... @XVilka @wargio WDYT?

XVilka replied:
UT32_MAX would be more appropriate, yes.

createyourpersonalaccount replied:
Okay so it returns the maximum value that fits in the return type. I will correct this.

ret2libc reviewed (commented):
Wow thanks for the documentation, it's great!

XVilka commented on librz/include/rz_endian.h:
```diff
@@ -646,24 +1581,192 @@ static inline void rz_write_ble(void *dst, ut64 val, bool big_endian, int size)
 static inline ut16 rz_swap_ut16(ut16 val) {
 	return (val << 8) | (val >> 8);
 }
+#define rz_swap_ut16 rz_swap_ut16
```
What is this for?

createyourpersonalaccount replied:
With Doxygen I document the `rz_swap_ut16` macro. I wasn't sure if the macro needed to be present or not, so I made sure it was always defined. Maybe I should just remove the extra `#define` and keep the Doxygen comment as it is?

XVilka replied:
Yes, just remove the macro

XVilka reviewed (changes requested):
If the function is wrong, please provide a fix, also update/create a test for it, see `test/unit/test_endian.c`
I also recommend checking the assembly output with [Compiler Explorer](https://godbolt.org/) for x86 and ARM64, to make sure the updated version produces fast and short code, it's in the "hot" path.

createyourpersonalaccount commented:
@XVilka Why not use this?

```c
st16 rz_swap_st16(st16 val) { return (st16)rz_swap_ut16((ut16)val); }
```

createyourpersonalaccount committed:
librz: remove UT24 macros

We remove the UT24_ macros due to PR review.

createyourpersonalaccount committed:
librz: replace UT24_MAX with UT32_MAX

It is better to return the maximum value held by the return type
instead of returning a max 24-bit value.

createyourpersonalaccount committed:
test: add more rz_endian tests

All the integral endian functions for read and write are ran through
some basic tests.

createyourpersonalaccount committed:
librz: remove unused internal functions

The signed rz_swap functions are removed because they are not used
anywhere in the source code. They are duplicates of the corresponding
unsigned functions. Moreover, the rz_swap_st16 function has an
incorrect implementation when the compiler builtin is missing.

Instead of fixing, adding documentation, and unit tests, which means
we would also need to create an environment with the compiler builtin
check undefined, we simply remove them, because they are not used
anywhere.

createyourpersonalaccount committed:
librz: remove unnecessary define macros

These macros were added during the PR and were requested to be removed
again.

createyourpersonalaccount commented:
@XVilka I decided to remove the signed swap functions, they were unused.

XVilka reviewed (changes requested):
With MSVC:
```
../test/unit/test_endian.c(428): error C2059: syntax error: ','
../test/unit/test_endian.c(429): error C2059: syntax error: ','
../test/unit/test_endian.c(430): error C2059: syntax error: ','
../test/unit/test_endian.c(431): error C2059: syntax error: ','
../test/unit/test_endian.c(432): error C2059: syntax error: ','
../test/unit/test_endian.c(434): error C2059: syntax error: ','
../test/unit/test_endian.c(435): error C2059: syntax error: ','
../test/unit/test_endian.c(436): error C2059: syntax error: ','
../test/unit/test_endian.c(437): error C2059: syntax error: ','
../test/unit/test_endian.c(438): error C2059: syntax error: ','
../test/unit/test_endian.c(440): error C2059: syntax error: ','
../test/unit/test_endian.c(441): error C2059: syntax error: ','
../test/unit/test_endian.c(442): error C2059: syntax error: ','
[1862/2031] Compiling C object test/unit/test_ebcdic.exe.p/test_ebcdic.c.obj
[1863/2031] Compiling C object test/unit/test_event.exe.p/test_event.c.obj
../test/unit/test_event.c(42): warning C4312: 'type cast': conve
```

createyourpersonalaccount committed:
test: fix MSVC complaining on erroneous commas

The string literal #__VA_ARGS__ is equal to "" in a conforming
implementation when there is an empty list of variadic arguments. On
MSVC 2017, it is completely missing, causing an error. These issues
appear to have been fixed in VS2019 v16.5 and later using the
/Zc:preprocessor compiler switch and in Visual Studio 2017 version
15.8 using /experimental:preprocessor.

In any case, we use string concatenating to force an empty string to
appear even if #__VA_ARGS__ is missing.

wargio reviewed (approved):
wow <3
